File: apps/frontend/packages/api-services/src/forms/application/search/SearchFormsQueryHandler.ts
import FormsSearcher from './FormsSearcher.ts';
import SearchFormsQuery from './SearchFormsQuery.ts';
import type { QueryHandler } from '@lyra/shared';
import type { PageResponse } from '~/types/types.ts';
import type { Form } from '~/forms/domain/Form.ts';
import FormResponse from '~/forms/application/FormResponse.ts';
import { inject, injectable } from 'inversify';
import { FORM_SEARCHER_PROVIDER } from '~/di/forms/forms.module.types.ts';

@injectable()
export default class SearchFormsQueryHandler
	implements QueryHandler<SearchFormsQuery, PageResponse<FormResponse>>
{
	constructor(@inject(FORM_SEARCHER_PROVIDER) private readonly formsSearcher: FormsSearcher) {}

	async handle(query: SearchFormsQuery): Promise<PageResponse<FormResponse>> {
		const pageResponse: Promise<PageResponse<Form>> = this.formsSearcher.search(query);

		const page = await pageResponse;
		return {
			...page,
			data: page.data.map((form: Form) => FormResponse.from(form)),
		};
	}
}

File: apps/backend/src/main/kotlin/com/lyra/app/forms/application/search/SearchFormsQueryHandler.kt
package com.lyra.app.forms.application.search

import com.lyra.app.forms.application.FormResponse
import com.lyra.common.domain.Service
import com.lyra.common.domain.bus.query.QueryHandler
import com.lyra.common.domain.presentation.pagination.CursorPageResponse
import org.slf4j.LoggerFactory

/**
 * This class represents a query handler to search all forms.
 *
 * @property searcher The [FormsSearcher] to search forms.
 */
@Service
class SearchFormsQueryHandler(
    private val searcher: FormsSearcher,
) : QueryHandler<SearchFormsQuery, CursorPageResponse<FormResponse>> {

    /**
     * Handles the given query.
     *
     * @param query The query to handle.
     * @return A CursorPageResponse containing the search results.
     */
    override suspend fun handle(query: SearchFormsQuery): CursorPageResponse<FormResponse> {
        log.info("Searching all forms")
        return searcher.search(query.criteria, query.size, query.cursor, query.sort)
    }
    companion object {
        private val log = LoggerFactory.getLogger(SearchFormsQueryHandler::class.java)
    }
}
